cowbird/permissions_synchronizer.py

```diff
@@ -74,59 +74,178 @@ def __init__(self,
+                         if res_key in self.resource_keys.keys()} for mapping_pt in mapping]
 
-    def resource_match(self, permission):
-        # type: (Permission) -> bool
-        """
-        Define if the permission name is covered by this sync point.
-        """
-        return permission.resource_full_name.startswith(self.services[permission.service_name])
-
-    def find_match(self, permission):
-        # type: (Permission) -> Generator[Tuple[str, str], None, None]
+    def find_permissions_to_sync(self, permission, res_root_key):
+        # type: (Permission, str) -> Generator[Tuple[str, str], None, None]
         """
         Search and yield for every match a (service, permission name) tuple that is mapped with this permission.
         """
-        # check if the permission name is covered by this sync point
-        if not self.resource_match(permission):
-            return
-
         # For each permission mapping
         for mapping in self.mapping:
             # Does the current service has some mapped permissions?
-            if permission.service_name not in mapping:
+            if res_root_key not in mapping:
                 continue
             # Does the current permission access is covered?
-            if permission.name not in mapping[permission.service_name]:
+            if permission.name not in mapping[res_root_key]:
                 continue
             # This permission is mapped : yields matches
-            for svc, mapped_perm_name in mapping.items():
+            for res, mapped_perm_name in mapping.items():
                 # Don't synch with itself
-                if svc == permission.service_name:
+                if res == res_root_key:
                     continue
                 for perm_name in mapped_perm_name:
-                    yield svc, perm_name
+                    yield res, perm_name
 
-    def sync(self, perm_operation, permission):
-        # type: (Callable[Permission], Permission) -> None
+    def _find_matching_res(self, service_name, resource_nametype_path):
+        # type: (str, str) -> (str, int)
+        """
+        Finds a resource key that matches the input resource path, in the sync_permissions config.
+        Note that it returns the longest match and only the named segments of the path are included in the length value.
+        Any tokenized segment is ignored in the length.
+
+        @param service_name Name of the service associated with the input resource.
+        @param resource_nametype_path Full resource path name, which includes the type of each segment
+                                       (ex.: /name1:type1/name2:type2)
+        """
+        service_resources = self.services[service_name]
+        # Find which resource from the config matches with the input permission's resource tree
+        # The length of a match is determined by the number of named segments matching the input resource.
```

Comment:
> I think it would be useful to have an example of two sample paths where match-length matters to determine what to do. It is not obvious to me which kind of situation could cause an ambiguity.

The case that I had thought about was with a config :
```
res1:
- name : **
  type: directory
- name : *
  type: file
res2:
- name : *
  type: file
```
A resource `file.txt` of type `file, would match with both. Considering the constraint that I use of only one possible match, it needs a way to know which one is better. I guess the one with the ** directory would be technically longer, but I am not sure if the code can really assume that in every case. 
I guess this example is an invalid config anyway, since if we ignore the length of the token, the 2 match would be of the same length, and it would give another error (2 matches of the same length, which is an ambiguous choice).

Another reason for avoiding tokens in the length is for example if we have this config :
```
res1:
- name : dir1
  type: directory
- name : file1
  type: file
res2:
- name : dir1
  type: directory
- name : *
  type: file
```
A path `dir1/file1` would match both resources. I think I would favor the one with the most specific names over a generic one, hence why I chose to ignore tokens in the length of a match.
This example would be useful if we need specific permissions for a specific file, but other permissions for the other files of the same folder. I have no idea if this case would happen, but it could be a possibility.

I tried to add those constraints since I wanted to avoid those specific cases.
We could always add more constraints on the config validation, and remove some constraints in the processing of the matching steps. 

Also, note that this part of the code is only to find the resource that matches with the input permission, the one defined in the received webhooks. Once the match is found, this resource can have any number of permissions to be synchronized in the mapping section of the config.

Reply:
> A resource `file.txt` of type `file`, would match with both. 

Are you sure about that? Since you add the `^...$` parts, it should match only on full paths, so only the `res1` case works in that example.

The only case I see where there would be dual full match would be if two definitions are equivalent, but one is more precise than the other, as in your second example, but the same could apply with `**` if one definition uses `/dir1/**/*` while another has `/dir1/dir2/dir3/**/*`.

So I guess ignoring the tokens is a good approach. Maybe just adding this example explicitly will help highlight why this length check is needed / implemented this way.

Reply:
It's a good idea to add more of these examples to the doc, since we will probably forget why it was done this way if I don't write it. I will complete the documentation.

> Are you sure about that? Since you add the ^...$ parts, it should match only on full paths, so only the res1 case works in that example.

With the config : 

```
res1:
- name : **
  type: directory
- name : *
  type: file
res2:
- name : *
  type: file
```
The path `/file` of type `file` would give a full match with both regex. Since the ** token can have 0 occurence, the path would match with the single token * of type `file` on both resources of the config, and the ** token would just not be used (0 occurence).